src/librustc_mir/interpret/eval_context.rs

```diff
@@ -23,14 +23,19 @@ use super::{
     Immediate, MPlaceTy, Machine, MemPlace, Memory, OpTy, Operand, Place, PlaceTy,
     ScalarMaybeUndef, StackPopInfo,
 };
+use rustc::infer::canonical::OriginalQueryValues;
+use rustc::infer::InferCtxt;
+use rustc::traits::ObligationCause;
 
-pub struct InterpCx<'mir, 'tcx, M: Machine<'mir, 'tcx>> {
+pub struct InterpCx<'infcx, 'mir, 'tcx, M: Machine<'mir, 'tcx>> {
     /// Stores the `Machine` instance.
     pub machine: M,
 
     /// The results of the type checker, from rustc.
     pub tcx: TyCtxtAt<'tcx>,
 
+    pub(super) infcx: &'infcx InferCtxt<'infcx, 'tcx>,
```

Comment:
Could you add a doc-comment to this? Also, as someone who works on Miri but has no idea about inference, what do I have to know about this new stuff that's coming to not screw it up when changing things?^^

Reply:
I have added in doc-comment in #67717. 

The thing to be aware of is that, when calling queries with values that may contain inference variables, it must either be a canonical query or you must check if the value contains any inference variables. There is still quite a few places within `rustc_mir::interpret`, which doesn't yet handle inference variables at all, so don't be too concerned about this yet :smile:. I'm confident that inference variables can only get into `const_eval` with the const generics feature.

Reply:
The thing is, I have no idea what a "canonical query" is. ;)